### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by Recillah-Khamala**
### **OOP school library: set up associations**

In this milestone, I have-:

- Created a class Classroom with the following:
  - @label instance variable, initialized in the constructor.
  - Setter and getter for @label.
- Created the has-many/belongs-to relationship between Classroom and Student. Implemented the below:
  - Created the has-many side (a classroom has many students).
  - Created the belongs-to side (a student belongs to a classroom).
  - Ensured that adding a student to a classroom also sets the classroom for the student.
  - Ensured that setting the classroom for a student also adds it to the classroom's students.
- Created a class Book with the below:
  - @title and @author instance variables, initialized in the constructor.
  - Setters and getters for instance variables.
- Created a class Rental with the following:
  - @date instance variable, initialized in the constructor.
  - Setter and getter for @date.
- Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental.  Implemented the below:
  - Created the has-many side of Book and Rental (a book has many rentals).
  - Created the belongs-to side of Rental and Book (a rental belongs to a book).
  - Created the has-many side of Person and Rental (a person has many rentals).
  - Created the belongs-to side of Rental and Person (a rental belongs to a person).
  - Modified the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: src/book.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+require_relative './rental'
+
+# Create a class Book
+class Book
+  # Setters and getters for instance variables
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (rloterh) [inline src/book.rb]: - [ ] `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _modify (set)_ it. Kindly ensure to improve upon this aspect.

File: src/person.rb [hunk 2]
```diff
@@ -1,9 +1,11 @@
 require_relative './nameable'
+require_relative './capitalize_decorator'
+require_relative './trimmer_decorator'
 
 # Make sure that your Person class inherits from Nameable
 class Person < Nameable
-  attr_accessor :name, :age, :parent_permission
-  attr_reader :id
+  attr_accessor :name, :age
+  attr_reader :id, :rentals
```
Comment (rloterh) [inline src/person.rb]: - [ ] Again, `rental` here should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _modify (set)_ it. Kindly ensure to improve upon this aspect.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Recillah, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- Nice job so far with creating associations among objects

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

**You can also consider:**
- Kindly ensure to add the _Future Features_ section to your ReadMe, which should include 1-3 potential features to be added.

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/book.rb (hunk 1 above)
Comment (Recillah-Khamala) [reply src/book.rb]: Change implemented

File: src/person.rb (hunk 2 above)
Comment (Recillah-Khamala) [reply src/person.rb]: Change implement

Comment (rloterh) [reply src/person.rb]: @Recillah-Khamala 
Kindly resubmit your project for a review

Cheers! 😃

Comment (Recillah-Khamala) [reply src/person.rb]: I have already submitted

File: src/classroom.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    @label = label
+    @students = []
+  end
+
+  #   Create the has-many side (a classroom has many students).
+  def add_student(student)
+    @students.push(student) unless @students.include?(student)
+    student.classroom = self
+  end
+end
```
Comment (tresorsawasawa) [inline src/classroom.rb:18]: - [ **OPTIONAL** ] In `README.md` **file:** As pointed out the previous reviewer, kindly make sure the you have added the ` Future Features` section in your README and this should include **1-3 potential features** to be added to your application.
Good :+1: job :100: adding other changes required by the previous reviewer.

**Review by tresorsawasawa — APPROVED** [review]
# STATUS: APPROVE :heavy_check_mark: :heavy_check_mark:  :champagne: 

Hi @Recillah-Khamala 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 :clap: :clap: :100: 

nodding-yes

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- Good PR description :heavy_check_mark: 
- Follow Gitflow :heavy_check_mark: 

## Optional suggestions
_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/classroom.rb (hunk 3 above)
Comment (Recillah-Khamala) [reply src/classroom.rb:18]: I have these in the features section. What the project would look like at the end of the project,

Comment (Recillah-Khamala) [reply src/classroom.rb:18]: Unless it should be something different.

File: src/person.rb (hunk 2 above)
Comment (tresorsawasawa) [reply src/person.rb]: Hi @Recillah-Khamala 

yes, I can see you have the [Key Features](https://github.com/Recillah-Khamala/School-Library/tree/associations#key-features) sub-section, but not the `## Future Features`.
Screenshot from 2022-12-23 04-49-33

Thanks

**Review by emmiiorji — COMMENTED** [review]
Hi @Recillah-Khamala, 

# Status: Approved ✅

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Congrats

## Good Highlights✔️

- All linters are passing ✔️
- Good commit history ✔️
- Good PR description ✔️
- Follow Gitflow ✔️

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**